src/mongo_logger.ts

```diff
@@ -111,69 +165,293 @@ function parseSeverityFromString(s?: string): SeverityLevel | null {
     return mongodbLogPath;
   }
 
-  if (typeof MONGODB_LOG_PATH === 'string' && isValidLogDestinationString(MONGODB_LOG_PATH)) {
-    return MONGODB_LOG_PATH.toLowerCase() === 'stderr' ? process.stderr : process.stdout;
+  if (MONGODB_LOG_PATH && /^stderr$/i.test(MONGODB_LOG_PATH)) {
+    return createStdioLogger(process.stderr);
+  }
+  if (MONGODB_LOG_PATH && /^stdout$/i.test(MONGODB_LOG_PATH)) {
+    return createStdioLogger(process.stdout);
   }
 
-  return process.stderr;
+  return createStdioLogger(process.stderr);
+}
+
+/** @internal */
+export interface Log extends Record<string, any> {
+  t: Date;
+  c: MongoLoggableComponent;
+  s: SeverityLevel;
+  message?: string;
+}
+
+/** @internal */
+export interface MongoDBLogWritable {
+  write(log: Log): void;
+}
+
+function compareSeverity(s0: SeverityLevel, s1: SeverityLevel): 1 | 0 | -1 {
+  const s0Num = SEVERITY_LEVEL_MAP.getNumericSeverityLevel(s0);
+  const s1Num = SEVERITY_LEVEL_MAP.getNumericSeverityLevel(s1);
+
+  return s0Num < s1Num ? -1 : s0Num > s1Num ? 1 : 0;
+}
+
+/** @internal */
+export type LoggableEvent =
+  | CommandStartedEvent
+  | CommandSucceededEvent
+  | CommandFailedEvent
+  | ConnectionPoolCreatedEvent
+  | ConnectionPoolReadyEvent
+  | ConnectionPoolClosedEvent
+  | ConnectionPoolClearedEvent
+  | ConnectionCreatedEvent
+  | ConnectionReadyEvent
+  | ConnectionClosedEvent
+  | ConnectionCheckedInEvent
+  | ConnectionCheckedOutEvent
+  | ConnectionCheckOutStartedEvent
+  | ConnectionCheckOutFailedEvent;
+
+/** @internal */
+export interface LogConvertible extends Record<string, any> {
+  toLog(): Record<string, any>;
+}
+
+/** @internal */
+export type Loggable = LoggableEvent | LogConvertible;
+
+function isLogConvertible(obj: Loggable): obj is LogConvertible {
+  const objAsLogConvertible = obj as LogConvertible;
+  // eslint-disable-next-line no-restricted-syntax
+  return objAsLogConvertible.toLog !== undefined && typeof objAsLogConvertible.toLog === 'function';
+}
+
+function attachCommandFields(
+  log: Record<string, any>,
+  commandEvent: CommandStartedEvent | CommandSucceededEvent | CommandFailedEvent
+) {
+  log.commandName = commandEvent.commandName;
+  log.requestId = commandEvent.requestId;
+  log.driverConnectionId = commandEvent?.connectionId;
+  const { host, port } = HostAddress.fromString(commandEvent.address).toHostPort();
+  log.serverHost = host;
+  log.serverPort = port;
+  if (commandEvent?.serviceId) {
+    log.serviceId = commandEvent.serviceId.toHexString();
+  }
+
+  return log;
+}
+
+function attachConnectionFields(
+  log: Record<string, any>,
+  connectionPoolEvent: ConnectionPoolMonitoringEvent
+) {
+  const { host, port } = HostAddress.fromString(connectionPoolEvent.address).toHostPort();
+  log.serverHost = host;
+  log.serverPort = port;
+
+  return log;
+}
+
+function defaultLogTransform(logObject: LoggableEvent): Omit<Log, 's' | 't' | 'c'> {
```

Comment:
Should this take more than just `LoggableEvent`? we seem to apply the transform to any object, so it may or may not be an actual event for our internal logging purposes.

Reply:
This is a good catch. We had set the type to `LoggableEvent` to allow for type narrowing and get editor completions but I forgot to expand the type to include arbitrary records.